src/node_sqlite.cc

```diff
@@ -128,6 +149,169 @@ inline void THROW_ERR_SQLITE_ERROR(Isolate* isolate, int errcode) {
+      e = CreateSQLiteError(isolate, pDest_).ToLocalChecked();
+
+      Cleanup();
+
+      resolver->Reject(env()->context(), e).ToChecked();
+
+      return;
+    }
+
+    pBackup_ = sqlite3_backup_init(
+        pDest_, dest_db_.c_str(), source_->Connection(), source_db_.c_str());
+
+    if (pBackup_ == nullptr) {
+      e = CreateSQLiteError(isolate, pDest_).ToLocalChecked();
+
+      sqlite3_close(pDest_);
+
+      resolver->Reject(env()->context(), e).ToChecked();
+
+      return;
+    }
+
+    this->ScheduleWork();
+  }
+
+  void DoThreadPoolWork() override {
+    backup_status_ = sqlite3_backup_step(pBackup_, pages_);
+  }
+
+  void AfterThreadPoolWork(int status) override {
+    HandleScope handle_scope(env()->isolate());
+
+    if (resolver_.IsEmpty()) {
+      Cleanup();
+
+      return;
+    }
+
+    Local<Promise::Resolver> resolver =
+        Local<Promise::Resolver>::New(env()->isolate(), resolver_);
+
+    if (!(backup_status_ == SQLITE_OK || backup_status_ == SQLITE_DONE ||
+          backup_status_ == SQLITE_BUSY || backup_status_ == SQLITE_LOCKED)) {
+      Local<Object> e =
+          CreateSQLiteError(env()->isolate(), backup_status_).ToLocalChecked();
+
+      Cleanup();
+
+      resolver->Reject(env()->context(), e).ToChecked();
+
+      return;
+    }
+
+    int total_pages = sqlite3_backup_pagecount(pBackup_);
+    int remaining_pages = sqlite3_backup_remaining(pBackup_);
+
+    if (remaining_pages != 0) {
+      Local<Function> fn =
+          Local<Function>::New(env()->isolate(), progressFunc_);
+
+      if (!fn.IsEmpty()) {
+        Local<Value> argv[] = {
+            Integer::New(env()->isolate(), total_pages),
+            Integer::New(env()->isolate(), remaining_pages),
+        };
+
+        TryCatch try_catch(env()->isolate());
+        fn->Call(env()->context(), Null(env()->isolate()), 2, argv)
+            .FromMaybe(Local<Value>());
+
+        if (try_catch.HasCaught()) {
+          Cleanup();
+
+          resolver->Reject(env()->context(), try_catch.Exception()).ToChecked();
+
+          return;
+        }
+      }
+
+      // There's still work to do
+      this->ScheduleWork();
+
+      return;
+    }
+
+    Local<String> message =
+        String::NewFromUtf8(
+            env()->isolate(), "Backup completed", NewStringType::kNormal)
+            .ToLocalChecked();
+
+    Local<Object> e =
+        CreateSQLiteError(env()->isolate(), pDest_).ToLocalChecked();
+
+    Cleanup();
+
+    if (backup_status_ == SQLITE_OK) {
+      resolver->Resolve(env()->context(), message).ToChecked();
+    } else {
+      resolver->Reject(env()->context(), e).ToChecked();
+    }
+  }
+
+  void Cleanup() {
+    if (pBackup_) {
+      sqlite3_backup_finish(pBackup_);
+      pBackup_ = nullptr;
+    }
+
+    if (pDest_) {
+      backup_status_ = sqlite3_errcode(pDest_);
+      sqlite3_close(pDest_);
+      pDest_ = nullptr;
+    }
+  }
+
+ private:
+  int backup_status_;
+  Environment* env() const { return env_; }
```

Comment:
We'll typically try to avoid mixing member methods and member vars. Best to separate those.

```suggestion
  Environment* env() const { return env_; }
  
  int backup_status_;
```